File: frontend/src/api/learning.ts
/** 学习中心 API。
 *
 * 与后端 `backend/app/api/v1/learning.py` 一一对应。
 * 判定在后端完成（AST 静态分析 + 真实数据探针），前端只负责展示与交互，
 * 因此这里没有「本地猜一个结果」的逻辑——所有结论都来自服务端。
 */
import { client, unwrap } from "./client";

export type TrackId = "ml" | "llm";
export type Level = "基础" | "进阶" | "挑战";
export type ProgressStatus = "not_started" | "in_progress" | "completed";

/** 单条检查结果。 */
export interface LearningCheck {
  requirement_index: number;
  kind: "ast" | "data";
  label: string;
  passed: boolean;
  detail: string;
  /** false 表示「未经校验」（如没选数据集时的数据检查），前端应显示为「未验证」而非「已通过」。 */
  verified: boolean;
}

/** 一次检查的完整结果。 */
export interface LearningReview {
  passed: boolean;
  score: number;
  checks: LearningCheck[];
  next_steps: string[];
  /** 语法错误等一票否决原因，有值时其他检查结果无意义。 */
  blocked: string | null;
  /** 数据集读取失败时的提示（结构检查仍已完成）。 */
  dataset_warning?: string;
  progress?: LearningProgressDetail;
}

export interface LearningProgressSummary {
  status: ProgressStatus;
  attempts: number;
  score: number;
  last_passed: boolean | null;
  completed_at: string | null;
  saved_at: string | null;
}

export interface LearningAttempt {
  score: number;
  passed: boolean;
  created_at: string | null;
  summary: string;
}

export interface LearningProgressDetail extends LearningProgressSummary {
  /** 已保存的代码；没有记录时后端返回 starter。 */
  code: string;
  dataset_id: number | null;
  /** 要求文本 -> 是否达成。 */
  checklist: Record<string, boolean>;
  last_result: Partial<LearningReview>;
  history: LearningAttempt[];
}

export interface LearningConcept {
  term: string;
  body: string;
}

export interface LearningReference {
  label: string;
  url: string;
}

export interface LearningExperiment {
  key: string;
  title: string;
  track: TrackId;
  level: Level;
  summary: string;
  why: string;
  goal: string;
  requirements: string[];
  starter: string;
  needs: string;
  minutes: number;
  prerequisite: string | null;
  concepts: LearningConcept[];
  references: LearningReference[];
  check_count: number;
  progress: LearningProgressSummary;
}

export interface LearningExperimentDetail extends Omit<LearningExperiment, "progress"> {
  progress: LearningProgressDetail;
  prerequisite_info?: {
    key: string;
    title: string;
    completed: boolean;
  };
}

export interface LearningTrack {
  id: TrackId;
  label: string;
  description: string;
  accent: string;
}

export interface LearningTrackStat {
  id: TrackId;
  label: string;
  total: number;
  completed: number;
  in_progress: number;
}

export interface LearningRecentItem {
  key: string;
  title: string;
  track: TrackId;
  status: ProgressStatus;
  score: number;
  attempts: number;
  updated_at: string | null;
}

export interface LearningOverview {
  total: number;
  started: number;
  completed: number;
  completion_rate: number;
  tracks: LearningTrackStat[];
  by_level: Record<string, number>;
  by_key: Record<string, { status: ProgressStatus; score: number; attempts: number }>;
  recommended: string | null;
  recent: LearningRecentItem[];
}

/** 数据体检面板的数据。 */
export interface LearningContextColumn {
  name: string;
  dtype: string;
  is_numeric: boolean;
  null_count: number;
  null_rate: number;
  unique_count: number;
}

export interface LearningContext {
  dataset_id: number;
  dataset_name: string;
  version: number;
  row_count: number;
  column_count: number;
  columns: LearningContextColumn[];
  numeric_columns: string[];
  label_candidates: string[];
  target_candidates: string[];
  missing_cells: number;
}

export interface CheckPayload {
  code: string;
  dataset_id?: number | null;
  version?: number | null;
  learner_id?: string;
  persist?: boolean;
}

/** 自建学习卡片。 */
export type CustomCardKind = "practice" | "checklist";

export interface CustomCardItem {
  text: string;
  done: boolean;
}

export interface CustomCardProgress {
  done: number;
  total: number;
  rate: number;
}

export interface CustomCard {
  id: number;
  learner_id: string;
  title: string;
  kind: CustomCardKind;
  goal: string;
  code: string;
  items: CustomCardItem[];
  progress: CustomCardProgress;
  created_at: string | null;
  updated_at: string | null;
}

export interface CustomCardCreate {
  title: string;
  kind: CustomCardKind;
  goal?: string;
  code?: string;
  items?: CustomCardItem[];
  learner_id?: string;
}

export interface CustomCardUpdate {
  title?: string;
  goal?: string;
  code?: string;
  items?: CustomCardItem[];
  learner_id?: string;
}

// tracks 是静态目录（两条主线），进程内缓存一次，避免每次进入学习中心都重新请求。
let _tracksCache: LearningTrack[] | null = null;

export function listTracks() {
  if (_tracksCache) return Promise.resolve(_tracksCache);
  return unwrap<LearningTrack[]>(client.get("/learning/tracks")).then((tracks) => {
    _tracksCache = tracks;
    return tracks;
  });
}

export function listExperiments(learnerId = "anonymous", track?: TrackId) {
  return unwrap<LearningExperiment[]>(
    client.get("/learning/experiments", {
      params: { learner_id: learnerId, track: track ?? undefined },
    }),
  );
}

export function getExperiment(key: string, learnerId = "anonymous") {
  return unwrap<LearningExperimentDetail>(
    client.get(`/learning/experiments/${key}`, {
      params: { learner_id: learnerId },
    }),
  );
}

export function checkSubmission(key: string, payload: CheckPayload) {
  return unwrap<LearningReview>(
    client.post(`/learning/experiments/${key}/check`, {
      code: payload.code,
      dataset_id: payload.dataset_id ?? null,
      version: payload.version ?? null,
      learner_id: payload.learner_id ?? "anonymous",
      persist: payload.persist ?? true,
    }),
  );
}

export function saveDraft(
  key: string,
  code: string,
  datasetId?: number | null,
  learnerId = "anonymous",
) {
  return unwrap<{ saved: boolean; saved_at: string | null; progress: LearningProgressSummary }>(
    client.post(`/learning/experiments/${key}/draft`, {
      code,
      dataset_id: datasetId ?? null,
      learner_id: learnerId,
    }),
  );
}

export function resetProgress(key: string, learnerId = "anonymous") {
  return unwrap<{ reset: boolean; key: string }>(
    client.delete(`/learning/experiments/${key}/progress`, {
      params: { learner_id: learnerId },
    }),
  );
}

export function getOverview(learnerId = "anonymous") {
  return unwrap<LearningOverview>(
    client.get("/learning/overview", { params: { learner_id: learnerId } }),
  );
}

export function getLearningContext(datasetId: number, version?: number) {
  return unwrap<LearningContext>(
    client.get(`/learning/datasets/${datasetId}/learning-context`, {
      params: { version: version ?? undefined },
    }),
  );
}

// ---------------------------------------------------------------------------
// 自建学习卡片（practice / checklist）
// ---------------------------------------------------------------------------

export function listCustomCards(learnerId = "anonymous") {
  return unwrap<CustomCard[]>(
    client.get("/learning/custom-cards", { params: { learner_id: learnerId } }),
  );
}

export function getCustomCard(cardId: number, learnerId = "anonymous") {
  return unwrap<CustomCard>(
    client.get(`/learning/custom-cards/${cardId}`, {
      params: { learner_id: learnerId },
    }),
  );
}

export function createCustomCard(payload: CustomCardCreate) {
  return unwrap<CustomCard>(
    client.post("/learning/custom-cards", {
      title: payload.title,
      kind: payload.kind,
      goal: payload.goal ?? "",
      code: payload.code ?? "",
      items: payload.items ?? [],
      learner_id: payload.learner_id ?? "anonymous",
    }),
  );
}

export function updateCustomCard(cardId: number, payload: CustomCardUpdate) {
  return unwrap<CustomCard>(
    client.patch(`/learning/custom-cards/${cardId}`, {
      title: payload.title,
      goal: payload.goal,
      code: payload.code,
      items: payload.items,
      learner_id: payload.learner_id ?? "anonymous",
    }),
  );
}

export function deleteCustomCard(cardId: number, learnerId = "anonymous") {
  return unwrap<{ deleted: boolean; card_id: number }>(
    client.delete(`/learning/custom-cards/${cardId}`, {
      params: { learner_id: learnerId },
    }),
  );
}

File: backend/app/api/v1/learning.py
"""学习中心 API。

端点一览：
    GET    /learning/tracks                  两条主线的定义
    GET    /learning/experiments             实验目录（含该学习者进度）
    GET    /learning/experiments/{key}       单个实验详情（含草稿与历史）
    POST   /learning/experiments/{key}/check 检查一次提交（可持久化）
    POST   /learning/experiments/{key}/draft 保存草稿
    DELETE /learning/experiments/{key}/progress  重置进度
    GET    /learning/overview                学习总览

设计要点：
- ``learner_id`` 是查询/表单参数，默认 anonymous。项目无鉴权（既定取舍），
  因此这里只做「学习者分区」而非「身份认证」，安全讨论写在论文局限里。
- 检查同时做 AST 静态分析与真实数据集探针，**不执行提交的代码**。
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field

from app.api.deps import get_dataset_service, get_learning_service
from app.core.exceptions import ValidationException
from app.learning.service import LearningService
from app.schemas.common import ApiResponse
from app.services.dataset_service import DatasetService

router = APIRouter(prefix="/learning", tags=["learning"])

# 学习场景下预览用的行列上限：够看清结构，又不至于把大表搬进内存
_CANDIDATE_COLUMNS = 40


class CheckRequest(BaseModel):
    code: str = Field(default="", description="学习者提交的代码")
    dataset_id: int | None = Field(default=None, description="用于数据侧校验与上下文")
    version: int | None = Field(default=None, description="数据版本（缺省最新）")
    learner_id: str = Field(default="anonymous")
    persist: bool = Field(
        default=True, description="是否把这次检查记入进度（临时预览可传 false）"
    )


class DraftRequest(BaseModel):
    code: str = Field(default="")
    dataset_id: int | None = None
    learner_id: str = Field(default="anonymous")


class CustomCardCreate(BaseModel):
    title: str = Field(default="", description="卡片标题")
    kind: str = Field(default="practice", description="practice / checklist")
    goal: str = Field(default="", description="目标描述（练习卡用于 AI 点评上下文）")
    code: str = Field(default="", description="起始代码（练习卡）")
    items: list[dict] = Field(default_factory=list, description="要点列表（清单卡）")
    learner_id: str = Field(default="anonymous")


class CustomCardUpdate(BaseModel):
    title: str | None = None
    goal: str | None = None
    code: str | None = None
    items: list[dict] | None = None
    learner_id: str = Field(default="anonymous")


@router.get("/tracks", response_model=ApiResponse[list])
def list_tracks() -> ApiResponse[list]:
    return ApiResponse[list](data=LearningService().list_tracks())


@router.get("/overview", response_model=ApiResponse[dict])
def overview(
    learner_id: str = Query("anonymous"),
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[dict]:
    """学习总览：完成度、主线分布、最近活动、推荐下一个实验。"""
    return ApiResponse[dict](data=service.overview(learner_id))


@router.get("/experiments", response_model=ApiResponse[list])
def list_experiments(
    learner_id: str = Query("anonymous"),
    track: str | None = Query(None, description="ml / llm，缺省返回全部"),
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[list]:
    return ApiResponse[list](
        data=service.list_experiments(learner_id, track=track)
    )


@router.get("/experiments/{key}", response_model=ApiResponse[dict])
def get_experiment(
    key: str,
    learner_id: str = Query("anonymous"),
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[dict]:
    return ApiResponse[dict](data=service.get_experiment(learner_id, key))


@router.post("/experiments/{key}/check", response_model=ApiResponse[dict])
def check_submission(
    key: str,
    body: CheckRequest,
    service: LearningService = Depends(get_learning_service),
    dataset_service: DatasetService = Depends(get_dataset_service),
) -> ApiResponse[dict]:
    """检查提交：AST 结构 + 真实数据探针，返回逐条要求的达成情况。

    这里加载真实数据集（而非让学习者上传），保证数据侧校验用的就是
    他界面上看到的那份，判定与所见一致。
    """
    if not body.code.strip():
        raise ValidationException("提交的代码为空")

    df = None
    dataset_error: str | None = None
    if body.dataset_id is not None:
        try:
            version_row = dataset_service.get_version_row(body.dataset_id, body.version)
            df = dataset_service.load_version(body.dataset_id, version_row.version)
        except Exception as exc:  # noqa: BLE001 - 数据不可用不应阻断结构检查
            dataset_error = str(exc)

    result = service.check(
        body.learner_id,
        key,
        body.code,
        df,
        dataset_id=body.dataset_id,
        persist=body.persist,
    )
    if dataset_error:
        result["dataset_warning"] = f"数据集读取失败，仅完成结构检查：{dataset_error}"
    return ApiResponse[dict](data=result)


@router.post("/experiments/{key}/draft", response_model=ApiResponse[dict])
def save_draft(
    key: str,
    body: DraftRequest,
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[dict]:
    """保存草稿：落代码与数据集，不改完成状态。"""
    return ApiResponse[dict](
        data=service.save_draft(
            body.learner_id, key, body.code, dataset_id=body.dataset_id
        )
    )


@router.delete("/experiments/{key}/progress", response_model=ApiResponse[dict])
def reset_progress(
    key: str,
    learner_id: str = Query("anonymous"),
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[dict]:
    return ApiResponse[dict](data=service.reset(learner_id, key))


@router.get("/datasets/{dataset_id}/learning-context", response_model=ApiResponse[dict])
def learning_context(
    dataset_id: int,
    version: int | None = Query(None),
    service: DatasetService = Depends(get_dataset_service),
) -> ApiResponse[dict]:
    """给学习台的「数据体检」面板：真实行数、列清单、缺失情况。

    学习场景需要先看懂数据再动手，这个接口把「看数据」从 AI 对话里
    拆出来，变成确定性的、可复核的展示。
    """
    version_row = service.get_version_row(dataset_id, version)
    dataset = service.get(dataset_id)
    df = service.load_version(dataset_id, version_row.version)

    columns: list[dict[str, Any]] = []
    for name, dtype in list(df.schema.items())[:_CANDIDATE_COLUMNS]:
        series = df[name]
        nulls = int(series.null_count())
        columns.append(
            {
                "name": name,
                "dtype": str(dtype),
                "is_numeric": bool(dtype.is_numeric()),
                "null_count": nulls,
                "null_rate": round(nulls / df.height, 4) if df.height else 0.0,
                "unique_count": int(series.n_unique()),
            }
        )

    numeric_columns = [c["name"] for c in columns if c["is_numeric"]]
    # 可作为分类标签的候选：基数低且非浮点
    label_candidates = [
        c["name"]
        for c in columns
        if not c["is_numeric"] and 1 < c["unique_count"] <= max(50, int(df.height * 0.05))
    ]
    # 可作为回归目标的候选：连续数值
    target_candidates = [
        c["name"]
        for c in columns
        if c["is_numeric"] and c["unique_count"] > max(10, int(df.height * 0.1))
    ]

    return ApiResponse[dict](
        data={
            "dataset_id": dataset_id,
            "dataset_name": dataset.name,
            "version": version_row.version,
            "row_count": df.height,
            "column_count": df.width,
            "columns": columns,
            "numeric_columns": numeric_columns,
            "label_candidates": label_candidates,
            "target_candidates": target_candidates,
            "missing_cells": int(df.null_count().sum_horizontal().item())
            if df.width
            else 0,
        }
    )


# ----------------------------------------------------------------------
# 自建学习卡片（practice / checklist）
# ----------------------------------------------------------------------


@router.get("/custom-cards", response_model=ApiResponse[list])
def list_custom_cards(
    learner_id: str = Query("anonymous"),
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[list]:
    return ApiResponse[list](data=service.list_custom_cards(learner_id))


@router.post("/custom-cards", response_model=ApiResponse[dict])
def create_custom_card(
    body: CustomCardCreate,
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[dict]:
    return ApiResponse[dict](
        data=service.create_custom_card(
            body.learner_id,
            title=body.title,
            kind=body.kind,
            goal=body.goal,
            code=body.code,
            items=body.items,
        )
    )


@router.get("/custom-cards/{card_id}", response_model=ApiResponse[dict])
def get_custom_card(
    card_id: int,
    learner_id: str = Query("anonymous"),
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[dict]:
    return ApiResponse[dict](data=service.get_custom_card(learner_id, card_id))


@router.patch("/custom-cards/{card_id}", response_model=ApiResponse[dict])
def update_custom_card(
    card_id: int,
    body: CustomCardUpdate,
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[dict]:
    return ApiResponse[dict](
        data=service.update_custom_card(
            body.learner_id,
            card_id,
            title=body.title,
            goal=body.goal,
            code=body.code,
            items=body.items,
        )
    )


@router.delete("/custom-cards/{card_id}", response_model=ApiResponse[dict])
def delete_custom_card(
    card_id: int,
    learner_id: str = Query("anonymous"),
    service: LearningService = Depends(get_learning_service),
) -> ApiResponse[dict]:
    return ApiResponse[dict](data=service.delete_custom_card(learner_id, card_id))
